allow NIP-44 to encrypt more than 65535 bytes

fiatjaf opened the pull request:
This is backwards-compatible and no one has to change anything if they do not care about bigger messages.
The only reasonable case where this necessity has shown up so far was signing of big kind:3 lists via NIP-46.

fiatjaf committed:
allow NIP-44 to encrypt more than 65535 bytes.

vitorpamplona commented:
There is no need for the 0 length signal. You can just check if the whole message is bigger than 87472 and go for the 4-byte length.

vitorpamplona commented:
Meaning.. current systems will misread the length bytes and will not be able to decrypt anyway. So why bother.

mikedilger commented on 44.md:
```diff
@@ -84,10 +84,12 @@ NIP-44 version 2 has the following design characteristics:
-   - Plaintext length is encoded in big-endian as first 2 bytes of the padded blob
+   - Plaintext length is encoded in big-endian:
+     - if smaller than 65536, as a u16 in the first 2 bytes of the padded blob;
+     - if greater than 65536, the first the first 6 bytes of the padded blob, the first 2 being zero and the other 4 being the actual encoded length as u32
```
the first the first

fiatjaf replied:
```suggestion
     - if greater than 65536, the first 6 bytes of the padded blob, the first 2 being zero and the other 4 being the actual encoded length as u32
```

fiatjaf committed:
fix typo

fiatjaf commented:
> There is no need for the 0 length signal. You can just check if the whole message is bigger than 65603 and go for the 4-byte length.

Sounds good to me. Is that number 65603 exact?

fiatjaf commented:
> We need to get some tests

But we'll go with my proposal or yours?

fiatjaf commented:
@vitorpamplona I think the current proposal is slightly easier to understand and the code for it looks clearer, and it's easier to debug for legacy implementations (you get a length=0, it raises some eyebrows).

So even if it wastes 2 bytes I think we should go with it. Yay or nay?

dannym-arx reviewed (approved):
Looks good to me!

fiatjaf commented:
Maybe we should just not allow people to have such big lists. 1800 is almost a hard limit of the protocol.

Most relays won't even accept events this big anyway. And no human can really "follow" that many people (sure, they can _click a button_ that many times, which isn't the same as really following).

aaccioly commented:
Agreed. I don’t think any of the points above are good reasons to keep NIP-44’s size limitations.

- ChaCha20 is a stream cipher, and it seems both the padding and the MAC can be handled on a block-by-block basis (correct me if I’m wrong). Despite the Python example code, as far as I can tell nothing in this algorithm requires bit arrays or holding more than a small buffer in memory for encryption or decryption (again, correct me if I’m wrong).
- We’ve already seen practical cases of people exceeding 64 KB follow lists, with NIP-44 breaking Nostr functionality as a result. On social media, "following" often equates to "connecting", and I personally had more than 1.8k connections on multiple platforms. Yes, we need to review how the Outbox model should work at scale, but that’s a separate problem. Also, this is a non-argument since NIP-44 is not used exclusively for follow lists (e.g., all other NIP-51 lists and sets, NIP-17 messages, etc.).
- Nostr, as a protocol, shouldn’t stop me from encrypting a payload of whatever size I want from my own client to my own relay. Limitations of popular Nostr software, including relays and clients, shouldn’t dictate protocol specs or encryption mechanisms.
- While I agree that size limitations make things easier for developers (since they don’t have to define their own policies). But I don’t think convenience should drive encryption design. In a way, this is similar to saying “NAT is just convenient”. Nostr needs a proper way to encrypt payloads larger than 64 KB. I also wouldn’t worry too much about performance here. I've used ChaCha20 to encrypt and decrypt multiple near real-time video streams simultaneously on consumer-grade hardware without hitting CPU limits. Setting a more reasonable limit for payloads shouldn’t have a major performance impact on clients. Even raising the cap to somewhere between 256 KB and 1024 KB would be a huge improvement in usability.

aaccioly commented:
> That's not to say we can't find a middle ground here, but 4GB probably isn't it.

The question, of course, is what is reasonable for which use cases, and ultimately what the boundaries are for Nostr (the protocol). A 4 GB limit is certainly unreasonable for short notes, comments, and follow lists, which is what much of Nostr’s early-stage software is optimised for. And of course, we have Blossom for media. But I can imagine many use cases where Nostr, as a protocol, could serve as a means to discover, write to and share a huge stream of ChaCha-encrypted data, with ephemeral notes that may never touch today’s relays but could drive tomorrow’s killer applications for Nostr.

My point is that NIP-44 is about cryptography. We shouldn’t artificially limit a payload encryption scheme just because it is convenient for the “Twitter over Nostr” use case, even if that is what a lot of Nostr currently looks like. New types of relays and clients can define their own adequate limits for their intended use cases. And if we want to define hard limits as part of the standard, I would prefer that it be done on a kind by kind basis, so we at least have more reasonable input on what’s adequate for each use case.

mikedilger commented:
For nip44, 2^32 is fine.  I updated my implementation.

For nostr events, gossip limits websocket messages to 1 MiB, so that sets a practical limit there.  Tunstenite library limit is 64 MiB.  Long ago I had heard people with >2k followers were hitting a 128 kb limit somewhere.

aaccioly commented:
So, 3 implementations and looks like we sorta kinda have an agreement. Merge? :)

alexgleason committed:
nip44: allow encryption of payloads larger than 65535 bytes

Extend the v2 padding format with a backwards-compatible sentinel:
when the first 2 bytes of the length prefix are zero, the next 4
bytes encode the plaintext length as a big-endian u32. This raises
the maximum from 65535 bytes to 2^32-1 bytes without requiring a
version bump.

Fixes from nostr-protocol/nips#1907:
- Fix off-by-one: use >= 65536 (not > 65536) for the extended path,
  since u16 can only represent 0..65535
- Fix padding validation: use dynamic prefix_len (2 or 6) instead of
  hardcoded 2 in the unpad() size check
- Fix len(d) typo in decode_payload (should be len(data))
- Remove upper-bound size checks in decode_payload that would reject
  large payloads
- Add write_u32_be, read_uint16_be, read_uint32_be to function list
- Add extended_prefix_threshold constant
- Update size range comments for both small and large payload paths

alexgleason committed:
nip44: fix pseudocode bugs, comment arithmetic, add implementation guidance and test vectors

alexgleason committed:
nip44: reject non-canonical extended prefix in unpad() pseudocode

When the 6-byte extended prefix sentinel is detected, validate that
the decoded length is >= extended_prefix_threshold (65536). Without
this, the same plaintext could be encoded with either prefix format,
breaking strict canonicalization.

alexgleason committed:
Merge nip44-big-payloads into bigger-nip44

alexgleason commented:
I added my own commits to this branch (from #2270)

Things addressed:

- Fix off-by-one: use >= 65536 (not > 65536) for the extended path, since u16 can only represent 0..65535
- Fix padding validation: use dynamic prefix_len (2 or 6) instead of hardcoded 2 in the unpad() size check
- Fix len(d) typo in decode_payload (should be len(data))
- Remove upper-bound size checks in decode_payload that would reject large payloads
- Add write_u32_be, read_uint16_be, read_uint32_be to function list
- Add extended_prefix_threshold constant
- Update size range comments for both small and large payload paths

nostr-tools update: https://github.com/nbd-wtf/nostr-tools/pull/527

erskingardner commented:
Looks like this is merged/implemented in nostr-tools now. Feels like it's ready to merge here after we get a few more implementations? Or are there outstanding questions still about the approach?

arthurfranca commented:
I have a concern about nip44's memory usage that has become more apparent with large payload support.

It doesn't look like this is making incremental ~hashing~encryption possible. I mean, for large payloads, we put everything in memory at once and pass it to the nip44 function as a big string.

If I'm not missing some detail, nip44 could've been the go-to encryption method for everything Nostr instead of relying on addition methods such as [on nip17 with aes-gcm](https://github.com/nostr-protocol/nips/blob/master/17.md#:~:text=encryption%2Dalgorithm%3A%20Indicates%20the%20encryption%20algorithm%20used%20for%20encrypting%20the%20file.%20Supported%20algorithms%3A%20aes%2Dgcm.).

Specially for large payloads, what is missing is acting on bytes instead of string and something akin to sha256.update() but for incremental encryption. This is more evident now that we want to support big payloads.

edit: not a blocker, just a thought that maybe someone else also thinks would be good to address
edit2: would certainly need a new nip07 method, so ignore this and go on ^^

paulmillr commented:
@arthurfranca that should be easy to do, will  think of a way

paulmillr commented:
@arthurfranca i've checked and can't find any part of algorithm which makes it unfriendly to streaming / chunking:

- chacha20 is parallel / stream-friendly. You can decrypt / encrypt on fly
- so is MAC, just do sha256.create().update(chunk1).update(chunk2).digest()
- Base64 is trickier but can still be done
- Padding is very simple operation

So just needs to be coded with proper tests. Even aes-gcm is harder to make streamable.

dannym-arx commented:
> Base64 is trickier but can still be done

base64 is easy to stream decode too, just always make sure to take multiples of 4

Semisol reviewed (changes requested):
I do not think this is a good idea:
1. People will try to encrypt files with NIP-44. This is a horrible idea as NIP-44 is not designed for random access.
2. Trying to use disk as temporary swap space for this type of operation will cause very fast wear-out of user devices' SSDs.
